src/tools/http_upload.c

```diff
@@ -0,0 +1,339 @@
+/*
+ * http_upload.c
+ *
+ * Copyright (C) 2016 Dominik Heidler <dominik@heidler.eu>
+ *
+ * This file is part of Profanity.
+ *
+ * Profanity is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Profanity is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Profanity.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link the code of portions of this program with the OpenSSL library under
+ * certain conditions as described in each individual source file, and
+ * distribute linked combinations including the two.
+ *
+ * You must obey the GNU General Public License in all respects for all of the
+ * code used other than OpenSSL. If you modify file(s) with this exception, you
+ * may extend this exception to your version of the file(s), but you are not
+ * obligated to do so. If you do not wish to do so, delete this exception
+ * statement from your version. If you delete this exception statement from all
+ * source files in the program, then also delete it here.
+ *
+ */
+
+#define _GNU_SOURCE 1
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <curl/curl.h>
+#include <gio/gio.h>
+#include <pthread.h>
+#include <assert.h>
+
+#include "profanity.h"
+#include "ui/ui.h"
+#include "ui/window.h"
+#include "tools/http_upload.h"
+#include "event/client_events.h"
+#include "config/preferences.h"
+
+#define FALLBACK_MIMETYPE "application/octet-stream"
+#define FILE_HEADER_BYTES 512
+
+struct curl_data_t {
+    char *buffer;
+    size_t size;
+};
+
+
+static int
+_xferinfo(void *userdata, curl_off_t dltotal, curl_off_t dlnow, curl_off_t ultotal, curl_off_t ulnow)
+{
+    HTTPUpload *upload = (HTTPUpload *)userdata;
+
+    pthread_mutex_lock(&lock);
+
+    if (upload->cancel) {
+        pthread_mutex_unlock(&lock);
+        return 1;
+    }
+
+    if (upload->bytes_sent == ulnow) {
+        pthread_mutex_unlock(&lock);
+        return 0;
+    } else {
+        upload->bytes_sent = ulnow;
+    }
+
+    unsigned int ulperc = 0;
+    if (ultotal != 0) {
+        ulperc = (100 * ulnow) / ultotal;
+    }
+
+    char *msg;
+    if (asprintf(&msg, "Uploading '%s': %d%%", upload->filename, ulperc) == -1) {
+        msg = NULL;
```

Comment:
I think passing NULL to the `win_update_entry_message` could cause a segfault as it does a `strdup(msg)`, not sure what the right thing to do is, maybe have a fallback message.